### ISSUE TITLE
Shade Guava manually

### ISSUE THREAD

**Issue by leventov** [apache/incubator-druid#6942]
It's suggested to create a module called "guava" in Druid that copies the Guava source tree (without tests) and prohibit using the real Guava in Druid source code via Checkstyle.

Shading Guava automatically using a plugin is not possible because Druid's dependencies depend on older Guava versions that are binary-incompatible with the freshest Guava. Repackaging all such dependencies doesn't seem to be a feasible task.

One could say that copying source code of a library is bad because we won't receive updates and bug fixes that happen in the upstream, but we aren't receiving them currently anyway, using the ancient 16.0.1 version!

**Comment by gianm** [comment]
Switching to a manually "shaded" version might even make it easier to keep our Guava version up to date, if we do the copy with a script that is easy to run against new Guava sources. The script could potentially copy just the Guava files we know we need and do minimal modifications to them in order to get them to be in different packages.

I think it's probably a good idea. We have so many problems caused by the combination of the Guava team's historical attitude towards backwards compatibility and the reluctance of the Hadoop community to upgrade to a non-ancient version of Guava. Recently the Guava team has had a more pragmatic attitude towards backwards compatibility (they have promised to support non-Beta APIs indefinitely). But sadly, this doesn't help us much, given that we still need to be compatible with versions from the older era.

**Comment by drcrallen** [comment]
Why not just polish off the last things that make it incompatible with newer or older guava versions?

**Comment by gianm** [comment]
> Why not just polish off the last things that make it incompatible with newer or older guava versions?

Do you mean upgrading our bundled Guava to the latest version, and sticking to only using APIs that exist in both the latest version and in Guava 11?

**Comment by drcrallen** [comment]
Keep the bundled at what it is. But make all functions we actually call highly compatible. That way installations can, if desired, substitute in their own guava version.

**Comment by gianm** [comment]
If we only call highly compatible functions, why not update to the latest? Maybe some of those functions have 'better' implementations in the latest version.

**Comment by drcrallen** [comment]
Well put! yes

**Comment by gianm** [comment]
It sounds like @leventov's suggestion, then, is just a more extreme version of what you're suggesting: do copy-or-switch-to-alternate for _all_ Guava APIs, not just ones that have some known incompatibility.

I guess to me, it doesn't really matter which one we do, as long as it achieves the goal of making Guava dependency problems a thing of the past. The extreme version sounds like it's more likely to do that and would be easier to implement. Are there advantages for the less extreme version?

**Comment by drcrallen** [comment]
The less extreme version allows us to be more cognizant of what we do and do not force versions in deployments. For example:

https://github.com/google/guava/releases/tag/v23.6.1 has some vulnerability fixes.

If we shade in a version, it makes operators unable to track down if they also have the vulnerability, and makes them near powerless to change libraries if they do.

**Comment by drcrallen** [comment]
See https://github.com/apache/incubator-druid/pull/6948 for example

**Comment by leventov** [comment]
To me it seems that there is a problem with Charles's approach (in #6948 and #6815) (also we have `Closer` and maybe some other classes copied from Guava already): this is very ad-hoc, each class is copied independently. There is no list of them all. Each class should be updated in separation. The likely result is that we probably won't update them.

**Comment by drcrallen** [comment]
I started with classloader reflection approaches, but people asked instead that the classes be copied over. (See the multiple declarations of support for copying the code in https://github.com/apache/incubator-druid/pull/6815 ).

I couldn't find a strong enough argument to refute it, and having the code in the codebase that we critically rely on has proven pretty sustainable in the past.

**Comment by jihoonson** [comment]
@drcrallen I'm still not sure what are the benefits of the reflection approach for this issue. Would you elaborate more?

**Comment by drcrallen** [comment]
One of the challenges I've faced with shading is that it becomes very hard to figure out which version of a library is actually deployed in a production cluster. With reflection, the jars are kept separate and it is easier to debug exactly which version things are coming from. With shading you loose the version information, and can accidentally shade (or neglect to shade) things which should not have (or should have) been shaded.

In such scenarios tracking down which version of what libs are getting pulled into which dependencies at what point in the build and deploy process is a huge pain. It is much easier to go "Oh, I see version X.Y.Z and X.Q.R in the `lib` dir, those must have a conflict" and then work out how to make them play nice together.

**Comment by leventov** [comment]
@drcrallen could you please reply to https://github.com/apache/incubator-druid/issues/6942#issuecomment-459687733?

**Comment by drcrallen** [comment]
Re https://github.com/apache/incubator-druid/issues/6942#issuecomment-459687733 : I do not see the alternative here. Being "stuck" on an old copy of the code is no worse than being stuck on an old version of the lib.

Yes it is ad hoc because it shouldn't be normal practice. Some of the functionalities are superseded by updates to other core libraries (like in the JRE itself) and the longer term support ones should be favored. Some can be modified to use apis that have long term support from Guava (or similar libraries) and should probably do that. But for some, the only way to maintain long term support in druid is to do it ourselves. Either by straight up copying the code, or by providing some introspection mechanism.

If you are worried about tracking how often we do this, maybe a 3rdparty module would suffice? We would keep code that has this kind of long-term-support wrappers in there?

**Comment by leventov** [comment]
@drcrallen it seems that we don't understand each other here.

> I do not see the alternative here.

The alternative is suggested by this issue: to copy the whole source tree of Guava as a module in Druid.

**Comment by gianm** [comment]
It sounds like the approaches being proposed are,

@drcrallen's proposal, embodied in #6948, to keep ourselves at Guava 16.0.1, but endeavor, through reflection and some limited copying, to be compatible with a wide range of Guava versions that may get subbed in at runtime.

@leventov's proposal, to copy the entire, latest Guava source tree into Druid, like we've done for Closer and some utility functions in the past.

I feel like @leventov's approach has a certain appeal to it -- it's simple, and 'guaranteed to work'. If we did it, it should really be done in an automated fashion, so we can easily stay up to date. But @drcrallen's has appeal too. A wholesale copy of a relatively large project like Guava into Druid is a bit ugly, so @drcrallen's approach is more elegant in my eyes. And we don't have to maintain a system that automates copying Guava into Druid as Guava evolves. And it gives users more control over what Guava version they run. #6948 also has the advantage of already being coded up. And, if we use the Guava 11 and Guava 'latest' profiles in CI, we can make sure that we are _actually_ compatible, giving us that 'guaranteed to work' assurance.

For these reasons, my opinion is let's go with @drcrallen's proposal.

**Comment by leventov** [comment]
> And it gives users more control over what Guava version they run.

In the approach that I suggested, we *also* depending on Guava 16.0.1, letting users to substitute Guava for other dependencies and third-party code users combine Druid with in their private builds, but not the Druid code itself. I don't really imagine why users may want to substitute the use of latest Guava in Druid code itself with use of some older version, except that if there is some regression in the latest Guava source, but that should then be considered a bug in Druid and fixed as a normal bug (since we control the source). But this is a very, very unlikely situation.

@gianm, your analysis also misses the main point and the advantage of the approach that I suggested: we are able to use latest Guava features, such as MoreCollectors, and more. I had to explain this earlier in the thread.

**Comment by gianm** [comment]
> @gianm, your analysis also misses the main point and the advantage of the approach that I suggested: we are able to use latest Guava features, such as MoreCollectors, and more. I had to explain this earlier in the thread.

Well, that's a good point. Why not merge #6948 (which works & solves a real problem) and then potentially implement a source-copying script later?

At the time we do that - probably motivated by wanting to use some new Guava features - I think ideally it only copies certain files from Guava. It's simpler to copy the entire tree, but ugly, since it would involve relocating a lot of stuff we don't use. So for tidiness purposes it seems worth it to maintain a list of files we want to copy and a script that copies just those files from a particular, provided Guava version.

**Comment by leventov** [comment]
On the other hand, when somebody discovers a small utility in Guava that they would use in their code, but it's not copied yet, they may decide to get away without using that small utility because of the complexity (or not knowing the process) of copying a new class from Guava.

**Comment by himanshug** [comment]
I encountered an issue due to guava conflicts and discovered this. I did a prototype using shading that appears to achieve the same effect as copying all of guava source code as suggested in this proposal.

I will likely send a PR soon, but here is what I did...

1. create a new module, called "druid-shaded-guava" , that has only guava-0.16.1 as its dependency and shade it to create a new package `org.apache.druid.<guava packages>`
2. Make all other druid modules depend upon above module and use `org.apache.druid.<guava packages>` . Remove their dependency on guava.
3. Some Druid classes are left that declare dependency on guava-0.16.1 (and classes in guava directly) that use another library whose interface depends upon guava classes e.g. in druid-sql module, few calcite method calls require guava Multimap. This would be needed even if we were to not shade and copy all of guava source code.

This allows us to use any versions of guava in `druid-shaded-guava` as we see fit without affecting anything else that is using real guava-0.x.y.jar packaged with druid.

**Comment by egor-ryashin** [comment]
I've been recently reviewing the PR https://github.com/apache/druid/pull/10964
Seems good, but we still have other libs dependent on the old Guava, ie:
```
mvn dependency:tree
...
[INFO] +- com.google.inject:guice:jar:4.1.0:provided
[INFO] |  +- aopalliance:aopalliance:jar:1.0:provided
[INFO] |  \- com.google.guava:guava:jar:16.0.1:compile
```
I've actually been blocked by the issue while developing an extension with Google libraries that internally used newer Guava. And after merging this https://github.com/apache/druid/pull/10964 the Druid will still be packaged with Guava jar of version 16. I suppose https://github.com/apache/druid/pull/10964 is the first step and there's a plan to upgrade Guice and other libs dependent on Guava 16?
(BTW, right now to load an extension into Druid I use Java Reflection along with a different classloader which is very constraining.)

**Comment by himanshug** [comment]
> And after merging this #10964 the Druid will still be packaged with Guava jar of version 16.

@egor-ryashin it is explained in PR  #10964  description , ".... Some imports in cloudfiles-extensions and sql modules are reverted to use import com.google.common.X because their dependencies jclouds-core and calcite respectively mandate using real import com.google.common.X in various methods exposed by those libraries. But this is minimal. ..."

**Comment by egor-ryashin** [comment]
Yes, I understand it's intentional. I was just wondering how to solve another problem, I wonder how to add an extension to Druid with a different Guava and I cannot shade it because I depend on it transitively through a bunch of other libs (and shading those libs breaks them).

**Comment by himanshug** [comment]
>  I wonder how to add an extension to Druid with a different Guava

I think it is best to use druid's shaded guava classes in the extension code .. extension code should be treated pretty much like the core code in that respect.

**Comment by egor-ryashin** [comment]
> > I wonder how to add an extension to Druid with a different Guava
> 
> I think it is best to use druid's shaded guava classes in the extension code .. extension code should be treated pretty much like the core code in that respect.

Yes, unfortunately, that approach won't work with the extension that uses dependency that internally depends on some other Guava.

**Comment by himanshug** [comment]
> Unfortunately, that approach won't work with the extension that uses dependency that internally depends on some other Guava.

that is same as druid-core depending upon calcite that depends upon guava, that problem will always be there if we have two dependencies that both transitively want conflicting versions of guava. Hope is that those two things can work with same version of guava.
